slang-api/src/main/java/org/sonarsource/slang/impl/TreeMetaDataProvider.java

```diff
@@ -50,6 +51,44 @@ public TreeMetaDataProvider(List<Comment> comments, List<Token> tokens) {
     return sortedComments;
   }
 
+  public List<Token> allTokens() {
+    return sortedTokens;
+  }
+
+  public int indexOfFirstToken(TextRange textRange) {
+    return indexOfFirstElement(sortedTokens, textRange);
+  }
+
+  public Optional<Token> firstToken(TextRange textRange) {
+    int textRangeIndex = indexOfFirstElement(sortedTokens, textRange);
+    if (textRangeIndex == -1) {
+      return Optional.empty();
+    } else {
+      return Optional.of(sortedTokens.get(textRangeIndex));
+    }
+  }
+
+  public Optional<Token> previousToken(TextRange textRange) {
+    int textRangeIndex = indexOfFirstElement(sortedTokens, textRange);
+    if (textRangeIndex <= 0) {
+      return Optional.empty();
+    } else {
+      return Optional.of(sortedTokens.get(textRangeIndex - 1));
+    }
+  }
+
+  public void updateTokenType(Token token, Token.Type newType) {
```

Comment:
It's a pity to break immutability. I hope that some other solutions were considered.